packages/hardhat/contracts/Ethernauts.sol

```diff
@@ -226,11 +211,30 @@ contract Ethernauts is ERC721Enumerable, Ownable {
         uint tokenId = totalSupply();
 
         _mint(to, tokenId);
+
+        _tryGenerateRandomNumber();
     }
 
     function _mint(address to, uint tokenId) internal virtual override {
         require(totalSupply() < maxTokens, "No available supply");
 
         super._mint(to, tokenId);
     }
+
+    function _tryGenerateRandomNumber() private {
+        uint randomNumberIdx = _randomNumbers.length;
+
+        uint maxTokenIdInBatch = batchSize * (randomNumberIdx + 1) - 1;
+        if (totalSupply() < maxTokenIdInBatch) {
+            return;
+        }
+
+        // solhint-disable not-rely-on-time
+        uint randomNumber = uint256(
+            keccak256(abi.encodePacked(msg.sender, block.difficulty, block.timestamp, _randomNumbers.length))
```

Comment:
it's more gas efficient to use randomNumberIdx than call again _randomNumbers.length

Reply:
Agree that the cached value saves a read from storage

Reply:
Notice that all the variables are known at time of minting, the number is not random

Reply:
```suggestion
            keccak256(abi.encodePacked(msg.sender, block.difficulty, block.timestamp, randomNumberIdx))
```